20-H0ngJu

H0ngJu opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
<!-- 해결한 문제의 링크를 올려주세요. -->
[스타트와 링크](https://www.acmicpc.net/problem/14889)

## ✔️ 소요된 시간
<!-- 문제를 해결하는데 소요된 시간을 적어주세요. -->
1H

## ✨ 수도 코드
<!-- 내가 작성한 코드를 모르는 사람이 봐도 이해할 수 있도록 글로 쉽게 풀어서 설명해주세요. -->
![image](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/106425326/4e6608f2-e0f7-4147-a491-557c75987bd4)

쪼매 많이 늦었습ㄴ디ㅏ .. ㅎ.. 제송합니다

<br/>

**_문제요약_**
스타트 팀과 링ㅋ 팀으로 나눠서 두 팀의 능력치가 최소가 되도록 하자@!

<br/>

**_수도코드_**
1. 절반이 될 때까지 인원을 나눈다 -> 스타트 팀은 visited 표시
2. 절반이 되면, 스타트 팀과 링크 팀의 능력치를 각각 더해준다.
3. 최솟값을 갱신 시켜준다.

<br/>
<br/>
스타트 팀과 링크 팀을 나누는 로직에 대해서 고민을 쫌 오래했었고,

answer를 처음에 global로 선언해주지 않아서 계속 0이 반환되었습니다.


로직의 문제인가? 해서 한참을 봤는데 변수 scope문제더라고요..


**함수를 쓸 때 변수 scope에 주의하자 .. !** << 이번 PR의 깨달음입니다.

```python
import sys

def input() : return sys.stdin.readline().rstrip()

N = int(input())
arr = [list(map(int, input().split())) for _ in range(N)]
visited = [0] * N
answer = 999999999
 
def solution(len, idx):
    global answer
    
    if len == N//2:
        link = 0
        start = 0

        for i in range (N):
            for j in range (N):
                if visited[i] and visited[j]: # 방문한 반만 팀
                    start += arr[i][j]
                elif not visited[i] and not visited[j]: # 나머지 반 팀
                    link += arr[i][j]
        
        answer = min(answer, abs(start-link))
        return
    
    else: # 인원 나누기
        for i in range(idx, N):
            if not visited[i]:
                visited[i] = 1
                solution(len+1, i+1)
                visited[i] = 0
 
solution(0,0)
print(answer)
```


## 📚 새롭게 알게된 내용
<!-- 새롭게 알게된 내용이 있다면 작성 해주시고 출처를 남겨주세요. -->

9kyo-hwang reviewed (approved):
파이썬이라 가능한, 비효율의 끝판왕 코드... ㅋㅋ
```py
from itertools import combinations
input = open(0).readline

N = int(input())
table = [list(map(int, input().split())) for _ in range(N)]
members = list(range(N))

ans = float('inf')
for start_members in combinations(members, N // 2):
    start, link = 0, 0
    link_members = list(set(members) - set(start_members))
    
    for m1, m2 in combinations(start_members, 2):
        start += table[m1][m2]
        start += table[m2][m1]
    
    for m1, m2 in combinations(link_members, 2):
        link += table[m1][m2]
        link += table[m2][m1]
        
    ans = min(ans, abs(start - link))
    
print(ans)
```

![image](https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/49135176/26f95be6-2b5a-4c10-9667-2c2a396f4e68)

---

C++로는 비트마스킹 써서 풀어봤습니다 ㅋㅋ
원래는 permutation을 이용해서 풀었었는데, 속도가 확 빨라지네요...

<details><summary>C++ 코드</summary>
<p>

```cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main()
{
    cin.tie(nullptr)->sync_with_stdio(false);
    
    int N; cin >> N;
    vector<vector<int>> Table(N, vector(N, 0));
    
    for(auto& Row : Table)
    {
        for(auto& Col : Row)
        {
            cin >> Col;
        }
    }
    
    int Answer = 40001;
    vector<int> Start, Link;
    for(int Comb = (N >> 1); Comb < (1 << N); ++Comb)
    {
        if(__builtin_popcount(Comb) != (N >> 1)) continue;
        
        Start.clear();
        Link.clear();
        
        for(int i = 0; i < N; ++i)
        {
            Comb & (1 << i) ? Start.emplace_back(i) : Link.emplace_back(i);
        }
        
        int StartScore = 0, LinkScore = 0;
        for(int i = 0; i < (N >> 1); ++i)
        {
            for(int j = 0; j < (N >> 1); ++j)
            {
                if(i == j) continue;
                StartScore += Table[Start[i]][Start[j]];
                LinkScore += Table[Link[i]][Link[j]];
            }
        }
        
        Answer = min(Answer, abs(StartScore - LinkScore));
    }
    
    cout << Answer;
}
```

</p>
</details>

tgyuuAn reviewed (approved):
이 문제 아무리 봐도 완탐으로 밖에 풀 수 없을 것 같아서 DFS로 풀었습니다.

```python
import sys

def input(): return sys.stdin.readline().rstrip()

N = int(input())
board = [list(map(int,input().split())) for _ in range(N)]
visited = set()
answer = int(1e9)

def dfs(board, str_start, prev_idx, depth):
    global visited, answer

    if depth == N//2 and str_start not in visited:
        team_start = 0        
        team_link = 0
        str_link = ""
        start = set(map(int,str_start.rstrip().split()))
        
        for row in range(N):
            if row not in start: str_link += f"{row} "

            for col in range(N):
                 if row in start and col in start: team_start += board[row][col]
                 elif row not in start and col not in start: team_link += board[row][col]
        
        answer = min(answer, abs(team_start - team_link))
        visited.add(str_start)
        visited.add(str_link)
        return
    
    for idx in range(prev_idx+1,N):
        new_str_start = str_start + f"{idx} "
        dfs(board, new_str_start, idx, depth+1)

    return

dfs(board, "", 0, 0)
print(answer)
```

근데 궁금한 점이 있는데,

**N이 최대 20**이면 저도 반절만 탐색한다고 생각해서 10개를 뽑는걸 생각을 했거든요,

<br><br><br><br><br><br>

근데 그럴 경우에, **20 * 19 * 18 ... * 10**이 되어서 시간 복잡도 터지지 않나요 ...?

이걸 계속 생각하느라 머뭇거렸는데 아무리 생각해도 이거밖에 없어서 결국 제출하긴했는데 solved 처리되네요.

제가 놓친 부분이 있나요? 설명 부탁드립니다들...

원 큐에 맞추긴 했는데 찝찝하네요.

H0ngJu commented:
> 이 문제 아무리 봐도 완탐으로 밖에 풀 수 없을 것 같아서 DFS로 풀었습니다.
> 
> ```python
> import sys
> 
> def input(): return sys.stdin.readline().rstrip()
> 
> N = int(input())
> board = [list(map(int,input().split())) for _ in range(N)]
> visited = set()
> answer = int(1e9)
> 
> def dfs(board, str_start, prev_idx, depth):
>     global visited, answer
> 
>     if depth == N//2 and str_start not in visited:
>         team_start = 0        
>         team_link = 0
>         str_link = ""
>         start = set(map(int,str_start.rstrip().split()))
>         
>         for row in range(N):
>             if row not in start: str_link += f"{row} "
> 
>             for col in range(N):
>                  if row in start and col in start: team_start += board[row][col]
>                  elif row not in start and col not in start: team_link += board[row][col]
>         
>         answer = min(answer, abs(team_start - team_link))
>         visited.add(str_start)
>         visited.add(str_link)
>         return
>     
>     for idx in range(prev_idx+1,N):
>         new_str_start = str_start + f"{idx} "
>         dfs(board, new_str_start, idx, depth+1)
> 
>     return
> 
> dfs(board, "", 0, 0)
> print(answer)
> ```
> 
> 근데 궁금한 점이 있는데,
> 
> **N이 최대 20**이면 저도 반절만 탐색한다고 생각해서 10개를 뽑는걸 생각을 했거든요,
> 
> 근데 그럴 경우에, **20 * 19 * 18 ... * 10**이 되어서 시간 복잡도 터지지 않나요 ...?
> 
> 이걸 계속 생각하느라 머뭇거렸는데 아무리 생각해도 이거밖에 없어서 결국 제출하긴했는데 solved 처리되네요.
> 
> 제가 놓친 부분이 있나요? 설명 부탁드립니다들...
> 
> 원 큐에 맞추긴 했는데 찝찝하네요.

시간 복잡도가 터진다는게 제한시간 내에 통과가 안될 것 같다는 뜻일까요??

9kyo-hwang commented:
> 시간 복잡도가 터진다는게 제한시간 내에 통과가 안될 것 같다는 뜻일까요??

이 부분은 디스코드 질문-답변 채널에서 이전에 [답변](https://discord.com/channels/1203381153764089958/1248095922806591571) 드린 내용입니다 :)
